File: Program's_Contributed_By_Contributors/Java_Programs/Nqueens.java
import java.util.Scanner;
public class NQueens 
{
    public static void main(String[] args) 
    {
        Scanner sc=new Scanner(System.in);
        System.out.println("Enter the value of n");
        int n=sc.nextInt();
        char chessBoard[][]=new char[n][n];
        for(int i=0;i<n;i++)
            for(int j=0;j<n;j++)
                chessBoard[i][j]='-';
        if(solveNQueens(chessBoard,0,n)) solution(chessBoard,n);
        else System.out.println("No solution exists");
    }
    public static void solution(char chessBoard[][], int n)
    {
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<n;j++)
                System.out.print(" "+chessBoard[i][j]+" ");
            System.out.println();
        }
    }
    public static boolean isSafe(char chessBoard[][], int row, int column, int n)
    {
        int i,j;
        for(i=0;i<column;i++)
        {
            if(chessBoard[row][i]=='Q') return false;
        }
        for(i=row,j=column; i>=0 && j>=0;i--,j--)
        {
            if(chessBoard[i][j]=='Q') return false;
        }
        for(i=row,j=column; i<n && j>=0;i++,j--)
        {
            if(chessBoard[i][j]=='Q') return false;
        }
        return true;
    }
    public static boolean solveNQueens(char chessBoard[][], int column, int n)
    {
        if(column>=n) return true;
        for(int i=0;i<n;i++)
        {
            if(isSafe(chessBoard,i,column,n))
            {
                chessBoard[i][column]='Q';
                if(solveNQueens(chessBoard, column+1,n)) return true;
                chessBoard[i][column]='-';
            }
        }
        return false;
    }
}

File: cpp/nqueens.cpp
#include <iostream>
#include <vector>

using namespace std;

bool isSafe(vector<vector<int>>& board, int row, int col, int N) {
    // Check if there is a queen in the same column
    for (int i = 0; i < row; ++i) {
        if (board[i][col] == 1) {
            return false;
        }
    }

    // Check upper-left diagonal
    for (int i = row, j = col; i >= 0 && j >= 0; --i, --j) {
        if (board[i][j] == 1) {
            return false;
        }
    }

    // Check upper-right diagonal
    for (int i = row, j = col; i >= 0 && j < N; --i, ++j) {
        if (board[i][j] == 1) {
            return false;
        }
    }

    return true;
}

bool solveNQueens(vector<vector<int>>& board, int row, int N) {
    if (row == N) {
        // All queens are placed successfully
        return true;
    }

    for (int col = 0; col < N; ++col) {
        if (isSafe(board, row, col, N)) {
            board[row][col] = 1;  // Place the queen

            // Recur to place the rest of the queens
            if (solveNQueens(board, row + 1, N)) {
                return true;
            }

            // If placing the queen in board[row][col] doesn't lead to a solution,
            // then backtrack and try the next column
            board[row][col] = 0;  // Backtrack
        }
    }

    // If no queen can be placed in this row, return false
    return false;
}

int main() {
    int N;
    cout << "Enter the value of N: ";
    cin >> N;

    vector<vector<int>> board(N, vector<int>(N, 0));

    if (solveNQueens(board, 0, N)) {
        // Print the solution
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                cout << board[i][j] << " ";
            }
            cout << endl;
        }
    } else {
        cout << "No solution found." << endl;
    }

    return 0;
}
